File: leetcode/lcs.py
#!/usr/bin/env python3

from collections import defaultdict
from functools import cache
import sys
import unittest

# number: 1143
# section: dp (multi-d)
# difficulty: medium
# tags: string, dp

# constraints
# 1 <= text1.length, text2.length <= 1000
# text1 and text2 consist of only lowercase English characters.

# solution: Leetcode top-down recursive 2DP using functools
# complexity
# run-time: O(n*m)
# space: O(n*m)
def lcs1(text1, text2):
    @cache
    def dp(i, j):
        # base case
        if i == len(text1) or j == len(text2):
            return 0

        # recurrence relation
        # match
        if text1[i] == text2[j]:
            return 1 + dp(i+1, j+1)

        # no match
        return max(dp(i, j+1), dp(i+1, j))

    return dp(0, 0)

# solution: Leetcode bottom-up iterative 2DP
# complexity
# run-time: O(n*m)
# space: O(n)
def lcs2(text1, text2):
    if not text1 or not text2:
        return 0

    # init: add 1!
    dp = [[0] * (len(text2)+1) for _ in range(len(text1)+1)]

    # recurrence relation
    # [start, stop) by step:
    # go backwards from last to first inclusive!
    for i in range(len(text1)-1, -1, -1):
        for j in range(len(text2)-1, -1, -1):
            # match
            if text1[i] == text2[j]:
                # advance both
                dp[i][j] = 1 + dp[i+1][j+1]
            # no match
            else:
                # add longest
                dp[i][j] = max(dp[i][j+1], dp[i+1][j])

    # longest stored in 1st element!
    return dp[0][0]

# non-solution: dict
# complexity
# run-time: O(n)
# space: O(n)
def lcs_bad1(text1, text2):
    count1 = defaultdict(int)
    count2 = defaultdict(int)

    for c in text1:
        count1[c] += 1

    for c in text2:
        count2[c] += 1

    subseq = ''
    i = j = 0

    while i < len(text1) and j < len(text2):
        #print("subseq:{}".format(subseq))

        # match
        if text1[i] == text2[j]:
            subseq += text1[i]
            count1[text1[i]] -= 1
            count2[text2[j]] -= 1
            i += 1
            j += 1
            continue
        # no match, but both exist later
        elif text1[i] in count2 and text2[j] in count1:
            count2[text2[j]] -= 1
            j += 1

        if text1[i] not in count2:
            i += 1

        if text2[j] not in count1:
            j += 1

    return len(subseq)

# non-solution: sorting loses order :(
# complexity
# run-time: O(n*log n)
# space: O(1)
def lcs_bad2(text1, text2):

    # O(n log n)
    text1 = ''.join(sorted(text1))
    text2 = ''.join(sorted(text2))
    subseq = ''
    i = j = 0

    # O(n)
    while i < len(text1) and j < len(text2):
        if text1[i] == text2[j]:
            subseq += text1[i]
            i += 1
            j += 1
        elif text1[i] < text2[j]:
            i += 1
        else:
            j += 1

    return len(subseq)

class TestLongestSubseq(unittest.TestCase):

    def test_empty(self):
        text1 = ''
        text2 = ''
        self.assertFalse(lcs_bad1(text1, text2))
        self.assertFalse(lcs1(text1, text2))
        self.assertFalse(lcs2(text1, text2))

    def test1(self):
        text1 = "abcde"
        text2 = "ace"
        expected = 3
        self.assertEqual(expected, lcs_bad1(text1, text2))
        self.assertEqual(expected, lcs1(text1, text2))
        self.assertEqual(expected, lcs2(text1, text2))

    def test2(self):
        text1 = "abc"
        text2 = "abc"
        expected = 3
        self.assertEqual(expected, lcs_bad1(text1, text2))
        self.assertEqual(expected, lcs1(text1, text2))
        self.assertEqual(expected, lcs2(text1, text2))

    def test3(self):
        text1 = "abc"
        text2 = "def"
        expected = 0
        self.assertEqual(expected, lcs_bad1(text1, text2))
        self.assertEqual(expected, lcs1(text1, text2))
        self.assertEqual(expected, lcs2(text1, text2))

    def test4(self):
        text1 = "ezupkr"
        text2 = "ubmrapg"
        expected = 2
        self.assertEqual(expected, lcs_bad1(text1, text2))
        self.assertEqual(expected, lcs1(text1, text2))
        self.assertEqual(expected, lcs2(text1, text2))

    def test5(self):
        text1 = "oxcpqrsvwf"
        text2 = "shmtulqrypy"
        expected = 2
        # fails
        #self.assertEqual(expected, lcs_bad1(text1, text2))
        self.assertEqual(expected, lcs1(text1, text2))
        self.assertEqual(expected, lcs2(text1, text2))

if __name__ == '__main__':
    sys.exit(unittest.main())

File: leetcode/lcs.cpp
#include <map>
#include <string>
#include <tuple>

// number: 1143
// title: Longest Common Subsequence
// url: https://leetcode.com/problems/longest-common-subsequence/
// section: dp (multi-d)
// difficulty: medium
// tags: string, dp

// constraints
// 1 <= text1.length, text2.length <= 1000
// text1 and text2 consist of only lowercase English characters.

int dp(int i,
       int j,
       const std::string& text1,
       const std::string& text2,
       std::map<std::tuple<int, int>, int>& memo)
{
    // base case
    if (i == text1.size() || j == text2.size()) {
        return 0;
    }

    // check cache
    const auto tuple = std::make_tuple(i,j);
    const auto& it = memo.find(tuple);
    if (it != memo.end()) {
        return it->second;
    }

    // recurrence relation
    // match
    if (text1[i] == text2[j]) {
        memo[tuple] = 1 + dp(i+1, j+1, text1, text2, memo);
    // no match
    } else {
        memo[tuple] = std::max(dp(i, j+1, text1, text2, memo),
                               dp(i+1, j, text1, text2, memo));
    }

    return memo[tuple];
}

// solution: Leetcode top-down recursive 2DP using map
// complexity
// run-time: O(n*m)
// space: O(n*m)
int lcs1(const std::string& text1, const std::string& text2)
{
    std::map<std::tuple<int,int>, int> memo;

    return dp(0, 0, text1, text2, memo);
}

// solution: Leetcode bottom-up iterative 2DP
// complexity
// run-time: O(n*m)
// space: O(n)
int lcs2(const std::string& text1, const std::string& text2)
{
    if (text1.empty() || text2.empty()) {
        return 0;
    }

    // init: add 1!
    std::vector<std::vector<int>> dp(
        text1.size()+1, std::vector<int>(text2.size()+1));

    // go backwards from last to first inclusive!
    for (int i = text1.size()-1; i >= 0; --i) {
        for (int j = text2.size()-1; j >= 0; --j) {
            // recurrence relations
            if (text1[i] == text2[j]) {
                dp[i][j] = 1 + dp[i+1][j+1];
            } else {
                dp[i][j] = std::max(dp[i][j+1], dp[i+1][j]);
            }
        }
    }

    // longest subseq stored in 1st el
    return dp[0][0];
}

// TODO: add unit tests
